src/nix/store-repair.cc

```diff
@@ -19,8 +19,32 @@ struct CmdStoreRepair : StorePathsCommand
 
     void run(ref<Store> store, StorePaths && storePaths) override
     {
+        size_t errors = 0;
+
+        // Remove duplicates.
+        StorePathSet storePathsSet;
         for (auto & path : storePaths)
-            store->repairPath(path);
+            storePathsSet.insert(path);
+
+        for (auto & path : storePathsSet) {
```

Comment:
This should be bottom-up, ie dependencies first. That way when we do a "substitution fallback" build, we don't use broken inputs.